sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/BroadcastHashJoinTranslator.java

```diff
@@ -42,15 +41,24 @@
   @Override
   PCollection<KV<KeyT, OutputT>> translate(
       Join<LeftT, RightT, KeyT, OutputT> operator,
-      PCollection<KV<KeyT, LeftT>> left,
-      PCollection<KV<KeyT, RightT>> right) {
+      PCollection<LeftT> left,
+      PCollection<KV<KeyT, LeftT>> leftKeyed,
+      PCollection<RightT> right,
+      PCollection<KV<KeyT, RightT>> rightKeyed) {
+
     final AccumulatorProvider accumulators =
         new LazyAccumulatorProvider(AccumulatorProvider.of(left.getPipeline()));
+
+    // We use PViewsStore to prevent multiple views to the same input PCollection.
+    // And therefore multiple broadcasts of the same data.
+    final PViewsStore pViews =
+        left.getPipeline().getOptions().as(EuphoriaOptions.class).getPCollectionViewsStore();
+
     switch (operator.getType()) {
       case LEFT:
         final PCollectionView<Map<KeyT, Iterable<RightT>>> broadcastRight =
-            right.apply(View.asMultimap());
-        return left.apply(
+            pViews.computeViewAsMultimapIfAbsent(right, rightKeyed);
```

Comment:
I don't think comparing by pcollection is enough as we can have a different key extractor.

Reply:
Agree. But that brigs the problem of comparing lambdas for equality (again). Which we do not have good enough solution for.  So I will use `==` equality and try to highlight it in the docs.